store/bigtable/bigtable.go

```diff
@@ -0,0 +1,470 @@
+	// the beginning. We can't assume we know the chunkSpan so we can't just calculate the t0 >= (start - <some-predefined-number>). ChunkSpans may change
+	// over time. We effectively need all chunks with a t0 > start, as well as the last chunk with a t0 <= start.
+	// Bigtable doesnt allow us to fetch the most recent chunk older then start, so instead we just fetch from (start - maxChunkSpan) to enure we get the data. This
+	// will usually result in more data being fetched then is needed, but that is an acceptable tradeoff.
+	adjustedStart := start - uint32(s.cfg.MaxChunkSpan.Seconds())
+
+	// chunkspans are aligned with Month_sec.  If start is > startMonth then the T0 of the chunk that start is in will also be >= startMonth.
+	if adjustedStart < startMonth {
+		adjustedStart = startMonth
+	}
+	agg := "raw"
+	if key.Archive > 0 {
+		agg = key.Archive.String()
+	}
+	// filter the results to just the agg method (Eg raw, min_60, max_1800, etc..) and the timerange we want.
+	// we fetch all columnFamilies (which are the different TTLS).  Typically there will be only one columnFamily
+	// that has data, unless the TTL of the agg has changed.  In which case we want all columnFamilies anyway.
+	filter := bigtable.ChainFilters(
+		bigtable.ColumnFilter(agg),
+		bigtable.TimestampRangeFilterMicros(bigtable.Timestamp(int64(adjustedStart)*1e6), bigtable.Timestamp(int64(end)*1e6)),
+	)
+
+	firstRow := formatRowKey(key, startMonth)
+	lastRow := formatRowKey(key, endMonth+Month_sec) // bigtable.RowRange is start inclusive, end exclusive
+	rr := bigtable.NewRange(firstRow, lastRow)
+	var err error
+	rowCount := 0
+	pre = time.Now()
+	queryCtx, cancel := context.WithTimeout(ctx, s.cfg.ReadTimeout)
+	reqErr := s.tbl.ReadRows(queryCtx, rr, func(row bigtable.Row) bool {
+		rowCount++
+		chunks := 0
+		var itgen *chunk.IterGen
+		for _, items := range row {
+			for _, rItem := range items {
+				chunkSizeAtLoad.Value(len(rItem.Value))
+				if len(rItem.Value) < 2 {
+					log.Errorf("btStore: bigtable readRows error. %s", err)
+					err = errChunkTooSmall
+					return false
+				}
+				itgen, err = chunk.NewGen(rItem.Value, uint32(rItem.Timestamp/1e6))
+				if err != nil {
+					log.Errorf("btStore: unable to create chunk from bytes. %s", err)
+					return false
+				}
+				chunks++
+
+				// This function is called serially so we dont need synchronization around adding to
+				// itgens.
+				itgens = append(itgens, *itgen)
+			}
+		}
+		btblChunksPerRow.Value(chunks)
+
+		return true
+	}, bigtable.RowFilter(filter))
+	cancel()
+	btblRowsPerResponse.Value(rowCount)
+	btblGetExecDuration.Value(time.Since(pre))
+
+	// free a slot in the readQueue
+	<-s.readQueue
+
+	if reqErr != nil {
+		log.Errorf("btStore: bigtable readRows error. %s", reqErr)
+		err = reqErr
+	}
+	if err != nil {
+		tracing.Failure(span)
+		tracing.Error(span, err)
+		btblReadError.Inc()
+	}
+	// TODO: do we need to ensure that itgens is sorted by chunk T0?
```

Comment:
i was wondering the same thing.
callers of store definitely expect the store to return in order.
whether the data is in order depends on bigtable, specifically i was wondering what happens when TTL is changed (which could be either lower or higher sorted CF)